provider/src/iosMain/kotlin/at/asitplus/crypto/provider/os/IOSKeychainProvider.kt

```diff
@@ -247,33 +265,61 @@ object IOSKeychainProvider:  TPMSigningProviderI<iosSigner<*>, iosSignerConfigur
-        }
 
         val publicKeyBytes: ByteArray = memScoped {
-            val attr = cfDictionaryOf(
-                kSecAttrKeyType to kSecAttrKeyTypeEC,
-                kSecAttrKeySizeInBits to 256,
-                kSecAttrTokenID to kSecAttrTokenIDSecureEnclave,
-                kSecPrivateKeyAttrs to cfDictionaryOf(
-                    kSecAttrLabel to alias,
-                    kSecAttrIsPermanent to true,
-                    kSecAttrApplicationTag to KeychainTags.PRIVATE_KEYS
-                ),
-                kSecPublicKeyAttrs to cfDictionaryOf(
+            val availability = config.tpm.v.let { c-> when (c.availability) {
+                iosTPMConfiguration.Availability.ALWAYS -> if (c.allowBackup) kSecAttrAccessibleAlways else kSecAttrAccessibleAlwaysThisDeviceOnly
+                iosTPMConfiguration.Availability.AFTER_FIRST_UNLOCK -> if (c.allowBackup) kSecAttrAccessibleAfterFirstUnlock else kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
+                iosTPMConfiguration.Availability.WHILE_UNLOCKED -> if (c.allowBackup) kSecAttrAccessibleWhenUnlocked else kSecAttrAccessibleWhenUnlockedThisDeviceOnly
+            } }
+            val attr = createCFDictionary {
+                when (val alg = config._algSpecific.v) {
+                    is SigningKeyConfiguration.ECConfiguration -> {
+                        kSecAttrKeyType mapsTo kSecAttrKeyTypeEC
+                        kSecAttrKeySizeInBits mapsTo alg.curve.coordinateLength.bits.toInt()
+                    }
+                    is SigningKeyConfiguration.RSAConfiguration -> {
+                        kSecAttrKeyType mapsTo kSecAttrKeyTypeRSA
+                        kSecAttrKeySizeInBits mapsTo alg.bits
+                    }
+                }
+                if (config.tpm.v.requireSecureEnclave) {
+                    kSecAttrTokenID mapsTo kSecAttrTokenIDSecureEnclave
+                }
+                kSecPrivateKeyAttrs mapsTo createCFDictionary {
+                    kSecAttrLabel mapsTo alias
+                    kSecAttrIsPermanent mapsTo true
+                    kSecAttrApplicationTag mapsTo KeychainTags.PRIVATE_KEYS
+                    when (val factors = config.tpm.v.protection.v?.factors?.v) {
+                        null -> {
+                            kSecAttrAccessible mapsTo availability
+                        }
+                        else -> {
+                            kSecAttrAccessControl mapsTo corecall {
+                                SecAccessControlCreateWithFlags(
+                                    null, availability,
+                                    when {
+                                        (factors.biometry && factors.deviceLock) -> kSecAccessControlUserPresence
+                                        factors.biometry -> kSecAccessControlBiometryAny
+                                        else -> kSecAccessControlDevicePasscode
```

Comment:
what if i want other flags? what if the assumptions here don't hold for my use case?

Reply:
anyways, nested cfdictionaries are segfaulting (no clue why). When constructing them in swift, the segfault goes away. This was not triggered on the simulator, since the combination of flags caused it to not get accesses through key generation (beacuse kSecAccessControl was never evaluated)

Reply:
resolved by 2941621